features/42-cloud-function-log-isolation/api.md

```diff
@@ -0,0 +1,256 @@
+        "user_required": true
+      }
+    ] 
+    "op": [
+      { /* same as map[string]interface{} in registrationInfo.Lambdas */
+        /* TODO */
+      }
+    ] 
+    "timer": [
+      {
+        "name": "",
+        "spec": ""
+      }
+    ] 
+    "provider": [
+      {
+        "type": "",
+        "name": ""
+      }
+    ] 
+  }
+}
+```
+
+### Invoke cloud function
+
+When calling this action, Skygear Server creates a plugin request struct and
+pass it to the plugin transport. This action returns to the client with data
+received from the plugin transport. 
+
+Request:
+
+```json
+{
+  "action": "cloudfunc:invoke",
+  "include_logs": false,
+  "func": {
+    "kind": "hook",
+    "name": "hook_name",
+    "type": "note",  /* only for db hook */
+    "trigger": "before_save"  /* only for db hook */
+  }
+  "param": {
+    /* the param is specific to the cloud function */
+  }
+}
+```
+
+Response:
+
+```json
+{
+  "result": {
+    "response": {
+      /* the response is specific to the cloud function */
+    }
+    "log": "<base64 encoded string>"
+  	}
+}
+```
+
+## Client SDK Changes
+
+The client SDK will not be updated because the purpose of this proposal is for
+testing cloud function. Client SDK serves app development.
+
+## Skycli Changes
+
+Find what cloud functions can be invoked:
+
+```
+$ skycli cloud function list
+TYPE      NAME           RECORD TYPE    TRIGGER
+hook      add_info       note           before_save
+lambda    get_messages
+timer     collect_money
+```
+
+Invoking a cloud function without payload will open editing session with
+template or payload from previous invocation.
+
+```
+$ skycli cloud function invoke hook add_info --record-type note --trigger before_save
+```
+
+`--record-type` and `--trigger` is not required if the name is unique.
+
+An editing session is opened with the following content. Save and exit to
+invoke.
+
+```
+# Save to invoke. This line is ignored.
+{
+  "func": {
+    "kind": "hook",
+    "name": "add_info",
+    "type": "note",
+    "trigger": "before_save"
+  },
+  "params": {
+    "record": { /* ... */ },
+    "old_record": { /* ... */ }
+  }
+}
+```
+
+Will print the result to stdout:
+
+```
+{ /* ... */ }
+```
+
+Other options to invoke cloud function:
+
+```
+$ skycli help cloud function invoke
+Usage: skycli cloud function invoke [options]
+    --again  Run with payload from previous invocation
+    --log    Print log to stderr
+
+$ skycli cloud function invoke note.add_info --again
+{ /* ... */ }
+
+$ skycli cloud function invoke note.add_info --again --log
+{ /* ... */ } 
+[py-skygear] ...log...
+```
+
+## Log Isolation
+
+Log isolation refers to obtaining console log from invoked cloud function. Log
+is to be isolated from each invocation of cloud functions.
+
+Since log is printed to stdout, we need to run cloud function in separate
+process so that log for individual invocation can be collected. This requires
+some changes across server and plugin runtime.
```

Comment:
Do we need to handle the state of the plugins? The plugins have their own lifecycles, i.e. register (using `includeme`) and init (using `init` event).